File: src/test/java/ru/jerael/booktracker/backend/web/security/AuthTokenFilterTest.java
package ru.jerael.booktracker.backend.web.security;

import jakarta.servlet.FilterChain;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.security.core.context.SecurityContextHolder;
import ru.jerael.booktracker.backend.domain.model.auth.IdentityTokenClaims;
import ru.jerael.booktracker.backend.domain.model.auth.IdentityTokenType;
import ru.jerael.booktracker.backend.domain.service.token.AuthTokenService;
import java.time.Instant;
import java.util.UUID;
import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class AuthTokenFilterTest {

    @Mock
    private AuthTokenService authTokenService;

    @Mock
    private HttpServletRequest httpServletRequest;

    @Mock
    private HttpServletResponse httpServletResponse;

    @Mock
    private FilterChain filterChain;

    @InjectMocks
    private AuthTokenFilter filter;

    private final UUID userId = UUID.fromString("ee39af7a-a073-4473-878a-1aae34e98bb7");
    private final String token = "token";

    @BeforeEach
    void setUp() {
        SecurityContextHolder.clearContext();
    }

    @Test
    void doFilterInternal_WhenValidToken_ShouldSetAuthentication() throws Exception {
        IdentityTokenClaims claims = new IdentityTokenClaims(
            userId,
            IdentityTokenType.ACCESS,
            "issuer",
            Instant.now().minusSeconds(100),
            Instant.now().plusSeconds(1000)
        );
        when(httpServletRequest.getHeader(any())).thenReturn("Bearer " + token);
        when(authTokenService.authenticateToken(token, IdentityTokenType.ACCESS)).thenReturn(claims);

        filter.doFilterInternal(httpServletRequest, httpServletResponse, filterChain);

        assertNotNull(SecurityContextHolder.getContext().getAuthentication());
        assertEquals(userId, SecurityContextHolder.getContext().getAuthentication().getPrincipal());
    }

    @Test
    void doFilterInternal_WhenNoToken_ShouldNotSetAuthentication() throws Exception {
        filter.doFilterInternal(httpServletRequest, httpServletResponse, filterChain);

        assertNull(SecurityContextHolder.getContext().getAuthentication());
    }
}

File: src/test/kotlin/ru/jerael/booktracker/backend/web/security/AuthTokenFilterTest.kt
package ru.jerael.booktracker.backend.web.security

import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.junit5.MockKExtension
import io.mockk.mockk
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertNotNull
import org.junit.jupiter.api.assertNull
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.web.servlet.HandlerExceptionResolver
import ru.jerael.booktracker.backend.domain.service.token.AuthTokenService
import ru.jerael.booktracker.backend.factory.auth.AuthDomainFactory

@ExtendWith(MockKExtension::class)
class AuthTokenFilterTest {
    private val authTokenService = mockk<AuthTokenService>()
    private val httpServletRequest = mockk<HttpServletRequest>()
    private val httpServletResponse = mockk<HttpServletResponse>()
    private val filterChain = mockk<FilterChain>(relaxed = true)
    private val resolver = mockk<HandlerExceptionResolver>(relaxed = true)
    
    @InjectMockKs
    private lateinit var filter: AuthTokenFilter
    
    @BeforeEach
    fun setUp() {
        SecurityContextHolder.clearContext()
    }
    
    @Test
    fun `when token is valid, doFilterInternal should set Authentication`() {
        val claims = AuthDomainFactory.createIdentityTokenClaims()
        val token = "token"
        every { httpServletRequest.getHeader("Authorization") } returns "Bearer $token"
        every { authTokenService.authenticateToken(token, any()) } returns claims
        
        filter.doFilterInternal(httpServletRequest, httpServletResponse, filterChain)
        
        val auth = SecurityContextHolder.getContext().authentication
        assertNotNull(auth)
        assertEquals(claims.userId, auth.principal)
    }
    
    @Test
    fun `when token is not provided, doFilterInternal should not set Authentication`() {
        filter.doFilterInternal(httpServletRequest, httpServletResponse, filterChain)
        
        val auth = SecurityContextHolder.getContext().authentication
        assertNull(auth)
    }
}